Mage.Sets/src/mage/cards/a/AnzragsRampage.java

```diff
@@ -0,0 +1,183 @@
+// Copied from StructuralAssaultWatcher
+class AnzragsRampageWatcher extends Watcher {
+
+    private int artifactsDied;
+
+    AnzragsRampageWatcher() {
+        super(WatcherScope.GAME);
+    }
+
+    @Override
+    public void watch(GameEvent event, Game game) {
+        if (event.getType() == GameEvent.EventType.ZONE_CHANGE) {
+            ZoneChangeEvent zEvent = (ZoneChangeEvent) event;
+            if (zEvent.isDiesEvent() && zEvent.getTarget().isArtifact(game)) {
+                artifactsDied++;
+            }
+        }
+    }
+
+    @Override
+    public void reset() {
+        super.reset();
+        artifactsDied = 0;
+    }
+
+    public int getArtifactsDied() {
+        return artifactsDied;
+    }
+}
+
+enum AnzragsRampageValue implements DynamicValue {
+    instance;
+
+    @Override
+    public int calculate(Game game, Ability sourceAbility, Effect effect) {
+        AnzragsRampageWatcher watcher = game.getState().getWatcher(AnzragsRampageWatcher.class);
+        if (watcher == null) {
+            return 0;
+        }
+        return watcher.getArtifactsDied();
+    }
+
+    @Override
+    public AnzragsRampageValue copy() {
+        return instance;
+    }
+
+    @Override
+    public String getMessage() {
+        return "";
+    }
+}
+
+class AnzragsRampageEffect extends OneShotEffect {
+
+    AnzragsRampageEffect() {
+        super(Outcome.PutCreatureInPlay);
+        staticText = "exile the top X cards of your library, where X is the number of artifacts " +
+                "that were put into graveyards from the battlefield this turn. " +
+                "You may put a creature card exiled this way onto the battlefield. It gains haste. " +
+                "Return it to your hand at the beginning of the next end step";
+    }
+
+    private AnzragsRampageEffect(final AnzragsRampageEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller == null) {
+            return false;
+        }
+
+        Cards cards = new CardsImpl(controller.getLibrary()
+                .getTopCards(game, AnzragsRampageValue.instance.calculate(game, source, this)));
+        controller.moveCardsToExile(
+                cards.getCards(game), source, game, true,
+                CardUtil.getExileZoneId(game, source),
+                CardUtil.createObjectRealtedWindowTitle(source, game, null));
+
+        TargetCard targetCard = new TargetCardInExile(
+                0, 1, StaticFilters.FILTER_CARD_CREATURE, CardUtil.getExileZoneId(game, source));
+        targetCard.withNotTarget(true);
+        controller.choose(outcome, targetCard, source, game);
+        Card card = game.getCard(targetCard.getFirstTarget());
+        if (card == null) {
+            return true;
+        }
+
+        if (controller.moveCards(card, Zone.BATTLEFIELD, source, game)) {
```

Comment:
If you use code with multiple zones moving then make sure it works with MDFC cards and on replacement effects like commander return.

Possible bugs/checking while testing:
- second delayed part works after blink, possible fix: miss mor usage somewhere;
- second part works after zone replacement applied (commander move, another effect), possible fix: card zone check before second part call;
- second part does not works (same replacement effect - see above), possible fix: miss state.ProcessActions after move to exile/grave/hand zones (move to battlefield works fine);

Reply:
^ code must be fine here cause it uses move to battlefield and new permanent after it. Just keep in mind about possible problems.

Reply:
Okay, awesome. I had started testing scenarios you mentioned, and everything was good so far:

- If I put a creature onto the battlefield and flicker it, the return-to-hand effect doesn't bounce it.
- I cast my commander and put it on top of my library with Time Ebb. Then I cast Anzrag's Rampage, and chose to put them onto the battlefield. At my end step, the return-to-hand effect triggers and I can decide if I want them in hand or in the command zone.